File: java/classic/divide_and_conquer/Karatsuba.java
package classic.divide_and_conquer;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class Karatsuba {
    public String multiply(String num1, String num2) {
        int length = Math.max(num1.length(), num2.length());
        if (length <= 4) {
            return String.valueOf(Integer.parseInt(num1) * Integer.parseInt(num2));
        }

        if (length > num1.length()) {
            num1 = fillHead(num1, length - num1.length());
        }
        if (length > num2.length()) {
            num2 = fillHead(num2, length - num2.length());
        }

        int halfSize = length >> 1;
        String num1Head = num1.substring(0, halfSize);
        String num1Tail = num1.substring(halfSize);
        String num2Head = num2.substring(0, halfSize);
        String num2Tail = num2.substring(halfSize);
        String aux1 = multiply(num1Head, num2Head);
        String aux2 = multiply(num1Tail, num2Tail);
        return add(add(fillTail(aux1, (length - halfSize) << 1), fillTail(subtract(multiply(add(num1Head, num1Tail), add(num2Head, num2Tail)), add(aux1, aux2)), length - halfSize)), aux2);
    }

    private String add(String num1, String num2) {
        int[] values1 = stringToArrays(num1);
        int[] values2 = stringToArrays(num2);

        int[] result = new int[Math.max(num1.length(), num2.length()) + 1];
        Arrays.fill(result, 0);
        System.arraycopy(values1, 0, result, 0, values1.length);
        for (int i = 0; i < values2.length; ++i) {
            result[i] += values2[i];
            if (result[i] >= 10) {
                result[i] -= 10;
                ++result[i + 1];
            }
        }

        int index = values2.length;
        while (index < values1.length - 1 && result[index] >= 10) {
            result[index] -= 10;
            ++result[index + 1];
            ++index;
        }
        return arraysToString(result);
    }

    private String subtract(String num1, String num2) {
        int[] values1 = stringToArrays(num1);
        int[] values2 = stringToArrays(num2);
        for (int i = 0; i < values2.length; ++i) {
            values1[i] -= values2[i];
            if (values1[i] < 0) {
                values1[i] += 10;
                --values1[i + 1];
            }
        }

        int index = values2.length;
        while (index < values1.length - 1 && values1[index] < 0) {
            values1[index] += 10;
            --values1[index + 1];
            ++index;
        }
        return arraysToString(values1);
    }

    private String fillHead(String num, int length) {
        return String.format("%0" + length + "d", 0) + num;
    }

    private String fillTail(String num, int length) {
        return num + String.format("%0" + length + "d", 0);
    }

    private int[] stringToArrays(String num) {
        List<Integer> result = new ArrayList<>();
        for (char value : num.toCharArray()) {
            result.add(value - '0');
        }
        Collections.reverse(result);
        return result.stream().mapToInt(Integer::valueOf).toArray();
    }

    private String arraysToString(int[] values) {
        List<Integer> aux = Arrays.stream(values).boxed().collect(Collectors.toList());
        Collections.reverse(aux);
        int index = 0;
        while (index < aux.size() - 1 && 0 == aux.get(index)) {
            ++index;
        }

        StringBuffer result = new StringBuffer();
        for (int i = index; i < aux.size(); ++i) {
            result.append(aux.get(i));
        }
        return result.toString();
    }
}

File: py/classic/divide_and_conquer/Karatsuba.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class Karatsuba:
    def multiply(self, num1: str, num2: str) -> str:
        length = max(len(num1), len(num2))
        if length <= 4:
            return str(int(num1) * int(num2))

        if length > len(num1):
            num1 = self.__fillHead(num1, length - len(num1))
        if length > len(num2):
            num2 = self.__fillHead(num2, length - len(num2))

        halfSize = length >> 1
        num1Head, num1Tail, num2Head, num2Tail = num1[ : halfSize], num1[halfSize : ], num2[ : halfSize], num2[halfSize : ]
        aux1, aux2 = self.multiply(num1Head, num2Head), self.multiply(num1Tail, num2Tail)
        return self.__add(self.__add(self.__fillTail(aux1, (length - halfSize) << 1), self.__fillTail(self.__subtract(self.multiply(self.__add(num1Head, num1Tail), self.__add(num2Head, num2Tail)), self.__add(aux1, aux2)), length - halfSize)), aux2)

    def __add(self, num1: str, num2: str) -> str:
        values1, values2 = self.__stringToArrays(num1), self.__stringToArrays(num2)

        result = [0 for _ in range(max(len(values1), len(values2)) + 1)]
        result[ : len(values1)] = values1[ : ]
        for i, value in enumerate(values2):
            result[i] += value
            if result[i] >= 10:
                result[i] -= 10
                result[i + 1] += 1

        index = len(values2)
        while index < len(values1) - 1 and result[index] >= 10:
            result[index] -= 10
            result[index + 1] += 1
            index += 1
        return self.__arraysToString(result)

    def __subtract(self, num1: str, num2: str) -> str:
        values1, values2 = self.__stringToArrays(num1), self.__stringToArrays(num2)
        for i, value in enumerate(values2):
            values1[i] -= value
            if values1[i] < 0:
                values1[i] += 10
                values1[i + 1] -= 1

        index = len(values2)
        while index < len(values1) - 1 and values1[index] < 0:
            values1[index] += 10
            values1[index + 1] -= 1
            index += 1
        return self.__arraysToString(values1)

    def __fillHead(self, num: str, length: int) -> str:
        return ('0' * length) + num

    def __fillTail(self, num: str, length: int) -> str:
        return num + ('0' * length)

    def __stringToArrays(self, num: str) -> List[int]:
        result = []
        for value in num:
            result.append(int(value))
        result.reverse()
        return result

    def __arraysToString(self, values: List[int]) -> str:
        values.reverse()
        index = 0
        while index < len(values) - 1 and not values[index]:
            index += 1

        result = ''
        for value in values[index : ]:
            result += str(value)
        return result
